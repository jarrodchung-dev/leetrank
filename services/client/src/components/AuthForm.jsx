import React, { Component } from "react";
import { Redirect } from "react-router-dom";
import axios from "axios";

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      type: "",
      username: "",
      email: "",
      password: "",
      disabled: true
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  componentDidMount() {
    this.setState({ type: this.props.type });
    this.clearForm();
  }
  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  handleSubmit(event) {
    event.preventDefault();
    const type = this.state.type;
    const data = {
      email: this.state.email,
      password: this.state.password
    };
    if (type === "Register") {
      data.username = this.state.username;
    }
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${type.toLowerCase()}`;
    axios
      .post(url, data)
      .then((res) => {
        this.clearForm();
        this.props.loginUser(res.data.auth_token);
      })
      .catch((err) => console.log(err));

    if (type === "Login") {
      this.props.createMessage(
        "The username and password provided were incorrect",
        "danger"
      );
    }
    if (type === "Register") {
      this.props.createMessage("Sorry, that user already exists", "danger");
    }
  }
  clearForm() {
    this.setState({ username: "", email: "", password: "" });
  }
  render() {
    if (this.props.isAuthenticated) {
      return <Redirect to="/" />;
    }
    return (
      <div>
        {this.props.type === "Login" && <h1 className="title">Log In</h1>}
        {this.props.type === "Register" && <h1 className="title">Register</h1>}
        <form onSubmit={this.handleSubmit}>
          {this.props.type === "Register" && (
            <div className="field">
              <label className="label">Username</label>
              <div className="control">
                <input
                  name="username"
                  value={this.state.username}
                  className="input"
                  type="text"
                  onChange={this.handleChange}
                />
              </div>
            </div>
          )}
          <div className="field">
            <label className="label">Email</label>
            <div className="control">
              <input
                name="email"
                value={this.state.email}
                className="input"
                type="email"
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="field">
            <label className="label">Password</label>
            <div className="control">
              <input
                name="password"
                value={this.state.password}
                className="input"
                type="password"
                onChange={this.handleChange}
              />
            </div>
          </div>
          <button className="button" type="submit" disabled={!this.state.disabled}>
            Submit
          </button>
        </form>
      </div>
    );
  }
}

export default Form;
