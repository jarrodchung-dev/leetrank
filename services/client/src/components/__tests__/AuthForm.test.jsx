import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import AuthForm from "../AuthForm.jsx";

const forms = [
  {
    type: "Register",
    title: "Register",
    data: {
      username: "",
      email: "",
      password: ""
    }
  },
  {
    type: "Login",
    title: "Log In",
    data: {
      email: "",
      password: ""
    }
  }
];

forms.forEach((form) => {
  describe(`<${form.type} Form /> (not authenticated)`, () => {
    const component = (
      <AuthForm type={form.type} data={form.data} isAuthenticated={false} />
    );
    const wrapper = shallow(component);

    it("renders the title of the form correctly", () => {
      const title = wrapper.find("h1.title");
      expect(title.length).toBe(1);
      expect(title.get(0).props.children).toBe(form.title);
    });

    it("renders only the fields that were passed in as props", () => {
      const fields = wrapper.find(".field");
      expect(fields.length).toBe(Object.keys(form.data).length);
    });

    it("submits the form correctly", () => {
      wrapper.instance().handleSubmit = jest.fn();
      wrapper.update();
      const input = wrapper.find(`input[type="email"]`);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalledTimes(0);
      input.simulate("change", {
        target: {
          name: "email",
          value: "test_user@example.com"
        }
      });
      wrapper.find("form").simulate("submit", form.data);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalledWith(form.data);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalledTimes(1);
    });

    it("renders a snapshot successfully", () => {
      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe(`<${form.type} Form /> (authenticated)`, () => {
    const component = (
      <AuthForm type={form.type} data={form.data} isAuthenticated={true} />
    );
    const wrapper = shallow(component);

    it("redirects away from <AuthForm />", () => {
      expect(wrapper.find("Redirect")).toHaveLength(1);
    });
  });
});

/*


*/
