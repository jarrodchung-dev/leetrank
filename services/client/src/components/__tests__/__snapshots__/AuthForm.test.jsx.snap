// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login Form /> (not authenticated) renders a snapshot successfully 1`] = `
<div>
  <h1
    className="title"
  >
    Log In
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="email"
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="button"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`<Register Form /> (not authenticated) renders a snapshot successfully 1`] = `
<div>
  <h1
    className="title"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label"
      >
        Username
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="username"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="email"
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="button"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
